extension/source/background/KeyringController.ts

```diff
@@ -214,4 +277,93 @@ export default class KeyringController {
       {},
     );
   }
+
+  async signWalletAddress(
+    senderAddress: string,
+    signerPrivateKey: string,
+  ): Promise<[BigNumberish, BigNumberish]> {
+    const netCfg = getNetworkConfig(
+      await this.network.read(),
+      this.multiNetworkConfig,
+    );
+
+    const addressMessage = solidityPack(['address'], [senderAddress]);
+    const wallet = await BlsWalletWrapper.connect(
+      signerPrivateKey,
+      netCfg.addresses.verificationGateway,
+      await this.ethersProvider.read(),
+    );
+    return wallet.signMessage(addressMessage);
+  }
+
+  async recoverWallet(
+    recoveryWalletAddress: string,
+    recoverySaltHash: string,
+    signerWalletPrivateKey: string,
+  ): Promise<string> {
+    const network = await this.network.read();
+    const netCfg = getNetworkConfig(network, this.multiNetworkConfig);
+
+    // Create new private key for the wallet we are recovering to.
+    const newPrivateKey = generateRandomHex(256);
+
+    const addressSignature = await this.signWalletAddress(
+      recoveryWalletAddress,
+      newPrivateKey,
+    );
```

Comment:
@JohnGuilding good question

1. Yeah it is.  It's the smart contract address of the 'instant wallet'.  So the wallet we are recovering.  Signed by the newly generated private key that will become the new private key for the address we are recovering.

3. I may not be the best at explaining this but this is my understanding. This signature is basically how the trusted wallet (the one that was set in the recovery hash) says that this private key now has permission to act on behalf of the instant wallet address.  When the recovery happens this private key will create a new public key but instead of deterministically generating an address the public key will map to the existing instant wallet address.  This signature is a way of verifying and setting this PK as the PK to the old address without revealing the private key.